psl-clj/src/psl_clj/core.clj

```diff
@@ -0,0 +1,347 @@
+   [org.linqs.psl.database Queries]
+   [java.util HashSet]
+   ))
+
+;;; ================== Building PSL models ===========================
+         
+(defn AND                
+  "Return a PSL conjunction formula over given formulas."
+  [& args]                                
+  (Conjunction. (into-array Formula args)))
+                                          
+(defn OR
+  "Return a PSL disjunction formula over given formulas."
+  [& args]                                
+  (Disjunction. (into-array Formula args)))
+                                          
+(defn IMPL                                
+  "Return a PSL rule formula for the given body and head formulas."
+  [body head]
+  (Implication. body head))
+                   
+(defn NOT          
+  "Return a PSL negation formula over the given formula."
+  [formula]
+  (Negation. formula))                
+                     
+(defmacro add-predicate 
+  "Add the specified predicate to the model."
+  [model pred-name argmap]
+  `(do 
+     (.add ~model                     ; Add the predicate to the model
+           (java.util.HashMap. 
+            (assoc ~argmap "predicate" (name '~pred-name)))) ; Add 'predicate'
+     (defn ~pred-name [& ~'args]    ; A function returning a QueryAtom
+       (QueryAtom. 
+        (.getPredicate ~model (name '~pred-name)) ; Predicate from name
+        (into-array                               ; Array of Terms
+         Term 
+         (for [~'a ~'args]
+           (if (symbol? ~'a) (Variable. (name ~'a)) ~'a)))))))
+                                      
+(defmacro add-pred-constraint         
+  "Add a predicate constraint to the model."
+  [model predicate constraint-type cname]
+  `(do
+     (.addConstraint
+      ~model
+      ~constraint-type
+      {"on" (.getPredicate ~model (name '~predicate))
+       "name" ~cname})))
+                       
+(defn add-rule [model formula weight squared name]
+  "Add a compatibility rule to the model."        
+  (let [rk (WeightedLogicalRule. formula weight squared name)]
+    (.addRule model rk)))
+
+;;; =============== Other functions for using PSL ====================
+                          
+(defn close-db            
+  "Close Database, catching an IllegalStateException if it is already closed."
+  [database]
+  (try
+    (.close database)
+    (catch IllegalStateException e (str "caught exception: " (.getMessage e)))))
+                                                                               
+(defn close-open-dbs "Close all open databases and returns their count."      
+  [datastore]                                                            
+  (let [dbs (.getOpenDatabases datastore)
+        num (count dbs)]
+    (doseq [d dbs] (close-db d))
+    num))
+        
+(defn default-inference
+  "Return an app for MPE inference using the configuration."
+  [model database config-bundle]
+  (LazyMPEInference. model database config-bundle))
```

Comment:
Why default to lazy?

Reply:
I just tried with MPEInference and got the following error during inference: "PersistedAccessException Can only call getAtom() on persisted RandomVariableAtoms using a PersistedAtomManager."  I'm looking at the examples in psl-examples for comparison, and don't see what I'm doing wrong yet.

Reply:
I wasn't specifying the full world of targets, so that was probably my problem.